File: src/main/java/io/scriptor/machine/Device.java
package io.scriptor.machine;

import io.scriptor.fdt.BuilderContext;
import io.scriptor.fdt.NodeBuilder;
import org.jetbrains.annotations.NotNull;

import java.io.PrintStream;

public interface Device {

    @NotNull Machine machine();

    void dump(@NotNull PrintStream out);

    void reset();

    default void step() {
    }

    default void build(@NotNull BuilderContext<Device> context, final @NotNull NodeBuilder builder) {
    }
}

File: src/main/java/io/scriptor/machine/Device.kt
package io.scriptor.machine

import io.scriptor.fdt.BuilderContext
import io.scriptor.fdt.NodeBuilder
import java.io.PrintStream

interface Device {

    val machine: Machine

    fun dump(out: PrintStream)

    fun reset()

    fun step() {
    }

    fun build(context: BuilderContext<Device>, builder: NodeBuilder) {
    }

    fun close() {
    }
}
